src/mount-support.c

```diff
@@ -416,6 +241,257 @@ static void sc_setup_mount_profiles(const char *security_tag)
+	debug("performing operation: mount --make-unbindable %s", scratch_dir);
+	if (mount("none", scratch_dir, NULL, MS_UNBINDABLE, NULL) < 0) {
+		die("cannot perform operation: mount --make-unbindable %s",
+		    scratch_dir);
+	}
+	// Recursively bind mount desired root filesystem directory over of the
+	// scratch directory. This puts the initial content into the scratch space
+	// and serves as a foundation for all subsequent operations below.
+	//
+	// The mount is recursive because it can either be applied the root
+	// filesystem of a core system (aka all-snap) or the core snap on a classic
+	// system. In the former case we need recursive bind mounts to accurately
+	// replicate the state of the root filesystem into the scratch directory.
+	debug("performing operation: mount --rbind %s %s", config->rootfs_dir,
+	      scratch_dir);
+	if (mount(config->rootfs_dir, scratch_dir, NULL, MS_REC | MS_BIND, NULL)
+	    < 0) {
+		die("cannot perform operation: mount --rbind %s %s",
+		    config->rootfs_dir, scratch_dir);
+	}
+	// Make the scratch directory recursively private. Nothing done there will
+	// be shared with any peer group, This effectively detaches us from the
+	// original namespace and coupled with pivot_root below serves as the
+	// foundation of the mount sandbox.
+	debug("performing operation: mount --make-rslave %s", scratch_dir);
+	if (mount("none", scratch_dir, NULL, MS_REC | MS_SLAVE, NULL) < 0) {
+		die("cannot perform operation: mount --make-rslave %s",
+		    scratch_dir);
+	}
+	// Bind mount certain directories from the host filesystem to the scratch
+	// directory. Mount events will propagate in both into and out of the
+	// peer group. This way the running application can alter any global state
+	// visible on the host and in other snaps.
+	for (const char **path = config->bidirectional_mounts; *path != NULL;
+	     path++) {
+		must_snprintf(dst, sizeof dst, "%s/%s", scratch_dir, *path);
+		debug("performing operation: mount --rbind %s %s", *path, dst);
+		if (mount(*path, dst, NULL, MS_REC | MS_BIND, NULL) < 0) {
+			die("cannot perform operation: mount --rbind %s %s",
+			    *path, dst);
+		}
+	}
+	// Bind mount certain directories from the host filesystem to the scratch directory.
+	// Mount events will only propagate inwards to the namespace. This way the running
+	// application cannot alter any global state apart from that of its own snap.
+	for (const char **path = config->unidirectional_mounts; *path != NULL;
+	     path++) {
+		must_snprintf(dst, sizeof dst, "%s%s", scratch_dir, *path);
+		debug("performing operation: mount --rbind %s %s", *path, dst);
+		if (mount(*path, dst, NULL, MS_REC | MS_BIND, NULL) != 0) {
+			die("cannot perform operation: mount --rbind %s %s",
+			    *path, dst);
+		}
+		debug("performing operation: mount --make-rslave %s", dst);
+		if (mount("none", dst, NULL, MS_REC | MS_SLAVE, NULL) != 0) {
+			die("cannot perform operation: mount --make-rslave %s",
+			    dst);
+		}
+	}
```

Comment:
I'm a little worried that bidirectional_mounts and unidirectional_mounts are two different lists and that bidirectional is also processed before unidirectional. The concern comes from the possible need to set up a unidirectional mount before a bidirectional. This could be solved by combining them into one list of something like:

```
struct sc_mounts {
    const char *path;
    bool is_bidirectional;
};
```

This is only an observation and not a blocker for this PR.